File: AppDrawersWeb/ApiServer/ItemTypes.cs
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace AppDrawers.ApiServer
{
    public class ItemTypes
    {
        public static readonly ItemTypes CaptureClip = new ItemTypes("CaptureClip", string.Empty, string.Empty, 0);

        public static readonly ItemTypes Clip = new ItemTypes(
            "Clip",
            ".Clip",
            @"%SystemRoot%\System32\SHELL32.dll",
            260);

        public static readonly ItemTypes DateFormat = new ItemTypes(
            "DateFormat",
            ".DateFormat",
            @"%SystemRoot%\System32\SHELL32.dll",
            167);

        public static readonly ItemTypes Directory = new ItemTypes(
            "Directory",
            string.Empty,
            @"%SystemRoot%\System32\SHELL32.dll",
            3);

        public static readonly ItemTypes NewFolder = new ItemTypes("NewFolder", string.Empty, string.Empty, 0);
        public static readonly ItemTypes NewShortcut = new ItemTypes("NewShortcut", string.Empty, string.Empty, 0);
        public static readonly ItemTypes OpenFolder = new ItemTypes("OpenFolder", string.Empty, string.Empty, 0);
        public static readonly ItemTypes Shortcut = new ItemTypes("Shortcut", ".lnk", string.Empty, 0);

        private ItemTypes(string type, string extension, string defaultIcon, int defaultIconIndex)
        {
            this.Type = type;
            this.Extension = extension;
            this.DefaultIcon = defaultIcon;
            this.DefaultIconIndex = defaultIconIndex;
        }

        public static List<ItemTypes> Values { get; } = new List<ItemTypes>()
        {
            NewFolder,
            OpenFolder,
            Directory,
            Shortcut,
            Clip,
            DateFormat,
            CaptureClip,
            NewShortcut
        };

        public string DefaultIcon { get; } = string.Empty;
        public int DefaultIconIndex { get; } = 0;
        public string Extension { get; } = string.Empty;
        public string Type { get; } = string.Empty;

        public static Regex GetExtensionMatchRegEx()
        {
            var sb = new StringBuilder();

            foreach (var type in Values)
            {
                if (type.Extension.Length > 0)
                {
                    if (sb.Length > 0)
                    {
                        sb.Append("|");
                    }

                    sb.Append($@"^.*\{type.Extension}$");
                }
            }

            return new Regex(sb.ToString());
        }

        public static ItemTypes GetValueByExtension(string extension)
        {
            foreach (var type in Values)
            {
                if (type.Extension.Equals(extension, System.StringComparison.OrdinalIgnoreCase))
                {
                    return type;
                }
            }

            return null;
        }
    }
}

File: AppDrawersWeb/GUI/www/AppDrawers/modules/models/ItemTypes.mjs
"use strict";

// noinspection JSUnusedGlobalSymbols
export class ItemTypes {

    static CaptureClip = new ItemTypes("CaptureClip", "", "", 0);

    static Clip = new ItemTypes(
        "Clip",
        ".Clip",
        "%SystemRoot%\\System32\\SHELL32.dll",
        260);

    static DateFormat = new ItemTypes(
        "DateFormat",
        ".DateFormat",
        "%SystemRoot%\\System32\\SHELL32.dll",
        167);

    static Directory = new ItemTypes(
        "Directory",
        "",
        "%SystemRoot%\\System32\\SHELL32.dll",
        3);

    static NewFolder = new ItemTypes("NewFolder", "", "", 0);
    static NewShortcut = new ItemTypes("NewShortcut", "", "", 0);
    static OpenFolder = new ItemTypes("OpenFolder", "", "", 0);
    static Shortcut = new ItemTypes("Shortcut", ".lnk", "", 0);

    DefaultIcon = "";
    DefaultIconIndex = 0;
    Extension = "";
    Type = "";

    constructor(type = "", extension = "", defaultIcon = "", defaultIconIndex = 0) {
        this.Type = type;
        this.Extension = extension;
        this.DefaultIcon = defaultIcon;
        this.DefaultIconIndex = defaultIconIndex;
    }

    /**
     * Converts the current object into a basic JavaScript object.
     * @returns {{}} A JavaScript interpretation of the current object.
     */
    ConvertToObject() {
        return {
            DefaultIcon: this.DefaultIcon,
            DefaultIconIndex: this.DefaultIconIndex,
            Extension: this.Extension,
            Type: this.Type,
        }
    }

    /**
     * Inspects a JavaScript object to determine if the current object's keys are present in the object. This implies that the JavaScript object can be converted into an instance of the current object model.
     * @param obj
     * @returns {boolean}
     */
    static CanConvert(obj) {
        return "DefaultIcon" in obj &&
            "DefaultIconIndex" in obj &&
            "Extension" in obj &&
            "Type" in obj
    }

    /**
     * Converts a JavaScript object into an instance of the current object model.
     * @param {object} data The JavaScript object to convert.
     * @returns {ItemTypes} An instance of the current object model.
     */
    static ConvertFrom(data) {

        if (!ItemTypes.CanConvert(data)) {
            throw "Unrecognized ItemTypes JSON.";
        }

        let it = new ItemTypes();

        it.DefaultIcon = data.DefaultIcon;
        it.DefaultIconIndex = data.DefaultIconIndex;
        it.Extension = data.Extension;
        it.Type = data.Type;

        return it;
    }
}